Custom reward addresses for masternodes

domob1812 opened the pull request:
Currently, masternode rewards are paid to the address that holds the collateral.  In some situations, this is restrictive, and it would be useful to give masternode owners a choice about where the reward should be paid to (for instance, this will allow easier setup of masternode vaults for improved security).

This set of changes implements a protocol upgrade, which then allows masternodes to (optinally) configure an arbitrary script to which rewards should be paid.  Until the upgrade and until a particular masternode owner decides to explicitly opt into this feature, rewards will continue to be paid to the default script (matching the collateral address).

The protocol update itself will be activated together with the "staking vault fork" (#38), namely when 95% of the past blocks have version 5.  This ensures that it happens in a decentralised and coordinated fashion, but only when a large majority of the network is already running code that supports the new logic.

domob1812 committed:
Introduce rewardScript for masternodes.

This adds a new field, CScript rewardScript, to the masternode data
structures.  For now, we just add it, return it from the RPCs, and
set it to the script corresponding to the collateral pubkey on
creation.

We also serialise the new field as part of CMasternode, which is used
e.g. in the on-disk cache, but not the wire protocol (where
CMasternodeBroadcast is used instead).  Thus this change invalidates the
mncache format, but has no other effects on compatibility.

In the future, we will implement a protocol upgrade and also network
fork that will use this new field so masternodes can specify an arbitrary
script to which they want to receive their payments, rather than being
paid always to the collateral address.

domob1812 committed:
New wire protocol for reward script.

Define a new wire protocol version, and for peers that match the
version, send the reward script as explicit part of the masternode
broadcast messages.

Also, if the reward script is not the default value, include it in the
signature message of the broadcast (so that the field is authenticated
by the collateral signature).

Until the protocol version is bumped on the network (which is not yet
the case with this commit), masternodes with non-default reward script
will not be accepted as valid.

domob1812 committed:
Pay to reward script of masternodes.

Instead of using the collateral pubkey, pay to the declared reward
script of masternodes (and use that to check payments).

domob1812 committed:
Test custom reward address in mnoperation.py.

Use a custom reward address for one of the two masternodes in the
mnoperation.py regtest, verifying that this works.

domob1812 committed:
Activate new protocol with fork.

This defines a new time-activated "fork", Fork::CustomRewardAddresses.
For now, it is scheduled at timestamp 2000000000; that will be set to
the actual activation time once it is determined.

When this fork becomes active (i.e. the current best block's timestamp
goes beyond the activation time), we require peers to use the new protocol
version that supports custom reward addresses.  The fork has no other
(in particular, no consensus-level) consequences.

domob1812 commented:
Once #46 is merged (as well as #38), the last commit (activation of the protocol change) needs to be updated.

domob1812 commented:
This is mostly ready, except that once #38 is merged, we can drop the commit that introduces the `StakingVaults` fork dummy.

domob1812 commented:
#38 is merged now, so this is ready for review.

galpHub commented on divi/src/masternode.h:
```diff
@@ -274,6 +290,15 @@ class CMasternodeBroadcast : public CMasternode
         READWRITE(vin);
         READWRITE(addr);
         READWRITE(pubKeyCollateralAddress);
+        if (nVersion >= MN_REWARD_SCRIPT_VERSION) {
+            READWRITE(rewardScript);
+        } else {
+            if (ser_action.ForRead()) {
+                rewardScript = GetDefaultRewardScript();
+            } else if (rewardScript != GetDefaultRewardScript()) {
+                LogPrintf("WARNING: CMasternodeBroadcast - ignoring changed reward script for serialisation of %s\n", vin.prevout.ToString());
+            }
+        }
```
The new reward script field should be placed at the end of the variable definitions, as well as its serialization to avoid becoming a hard fork

domob1812 replied:
As discussed, I think this is fine.  But I will work on explicit regtests to ensure compatibility, and mark this as draft for now.

domob1812 committed:
Regtest for update backwards compatibility.

This adds a new regtest, custom_address_update.py.  The test runs
two masternodes, where we explicitly set the protocol version of
one of the nodes to the "old" version not supporting custom reward
scripts for masternodes.

With this test, we ensure that all the masternode networking works fine
between "old" and "new" nodes before the update is activated, at least
as closely as we can without running a real old node.

domob1812 committed:
Move internal functions into anonymous namespace.

Wrap some of the functions used only in main into an anonymous namespace;
some of them were already marked as "static", others were not (but they
were still not needed from anywhere else).

This makes it immediately clear that those functions are only used in the
context of main.

FindUndoPos had a declaration first and the definition later, but it is
fine to just move the definition to where the declaration was.  This
simplifies the code further (and is a pure move in the context of
this commit).

domob1812 committed:
Clean up some mn config code.

This is a bunch of related and straight-forward cleanups to the code
around masternode configuration:  Pass arguments as const& instead of
by value, mark functions as const, and move helper functions from
being static in a class and declared inside the header to just being
inside an anonymous namespace in the cpp file.

domob1812 committed:
Allow configuration of reward address.

Allow setting a non-default reward script for masternodes, by editing
the masternode.conf file explicitly.  The RPC setupmasternode will not
generate such configuration lines for now, and also "old" masternode.conf
files will remain valid (with the default script).

domob1812 commented:
Reopened as #83.
